akka-stream/src/main/scala/akka/stream/remote/scaladsl/SinkRef.scala

```diff
@@ -0,0 +1,199 @@
+import akka.annotation.InternalApi
+import akka.pattern.FutureRef
+import akka.stream._
+import akka.stream.impl.SinkModule
+import akka.stream.remote.impl.StreamRefsMaster
+import akka.stream.remote.{ RemoteStreamRefActorTerminatedException, StreamRefs }
+import akka.stream.scaladsl.Source
+import akka.stream.stage._
+import akka.util.{ ByteString, OptionVal }
+import org.reactivestreams.Subscriber
+
+import scala.concurrent.{ Future, Promise }
+
+object SinkRef {
+  def source[T](): Source[T, Future[SinkRef[T]]] =
+    Source.fromGraph(new SinkRefLocalImpl[T]()) // TODO settings?
+
+  def bulkTransferSource(port: Int = -1): Source[ByteString, SinkRef[ByteString]] = {
+    ???
+  }
+}
+
+///**
+// * INTERNAL API
+// */
+//@InternalApi private[akka] final class SinkRefModule[In](val attributes: Attributes, shape: SinkShape[In]) extends SinkModule[In, ActorRef](shape) {
+//
+//  override def create(context: MaterializationContext): (Subscriber[In], ActorRef) = {
+//    val system = ActorMaterializerHelper.downcast(context.materializer)
+//    val subscriberRef = system.actorOf(context.copy(islandName = ""), SinkRefActor.props)
+//    (akka.stream.actor.ActorSubscriber[In](subscriberRef), subscriberRef)
+//  }
+//
+//  override protected def newInstance(shape: SinkShape[In]): SinkModule[In, ActorRef] = new SinkRefModule[In](attributes, shape)
+//  override def withAttributes(attr: Attributes): SinkModule[In, ActorRef] = new SinkRefModule[In](attr, amendShape(attr))
+//}
+
+final class SinkRefLocalImpl[T] extends GraphStageWithMaterializedValue[SourceShape[T], Future[SinkRef[T]]] {
+  final val out: Outlet[T] = Outlet[T]("SinkRefSource.out")
+  override final def shape = SourceShape.of(out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+    val promise = Promise[SinkRef[T]]()
+
+    /*
+     * This stage can only handle a single "sender" (it does not merge values);
+     * The first that pushes is assumed the one we are to trust
+     */
+    val logic = new GraphStageLogic(shape) with StageLogging with OutHandler {
+
+      //      var receiverRef: ActorRef = _
+
+      private[this] var self: GraphStageLogic.StageActor = _
+
+      var localCumulativeDemand = 0L
+      var remotePartner: ActorRef = _
+
+      override def preStart(): Unit = {
+        self = getStageActor(initialReceive)
+        //        val master = StreamRefsMaster(ActorMaterializerHelper.downcast(materializer).system)
+        //        receiverRef = master.allocateReceiver(self.ref)
+        //        log.warning("Allocated receiver: {}", receiverRef)
+
+        promise.success(new SinkRef(self.ref))
+      }
+
+      override def onPull(): Unit = {
+        log.warning("Pullin")
+        localCumulativeDemand += 1
+        triggerDemand()
+      }
+
+      def triggerDemand(): Unit = if (remotePartner ne null) {
+        remotePartner.tell(StreamRefs.CumulativeDemand(localCumulativeDemand), self.ref)
```

Comment:
Can you explain how the cumulative demand would work? Is there something missing to make it work? Wasn't the idea to notice that a message was lost?

From what I can see if either demand or data is lost then it will never be resent and the connection will just be silent. There will never be more than one demand on the local side. If that's what you want to achieve you will need a manual buffer on the receiving end.

Reply:
The re-sending is not implemented yet, yes.
Yes there will be a manual buffer inside the receiving stage. I think it's likely worth it so we don;t end up requesting one-by-one all the time